### PR TITLE
Fix credit note sequence issue

### PR DESCRIPTION
**Pull request by guanhuan**

### CODE REVIEW

**Comment by guanhuan** [comment]
Ref: https://issues.civicrm.org/jira/browse/CRM-17235

Credit note id field will be populated with the next available sequence when a contribution is cancelled or refunded without "Credit note id" field manually filled in.

@davecivicrm Please take a look and let me know if this is okay to be merged. Thanks.

Best

Guanhuan

**Comment by davecivicrm** [comment]
@guanhuan This works, BUT you have the same lines of code repeated several times. Pretty sure this can be refactored a bit (create a new function in the BAO) - important for maintainability AND testability. Once you have the new BAO function, you can also add a simple unit test for it. Thanks!

**Comment by guanhuan** [comment]
jenkins, test this please

**Comment by guanhuan** [comment]
Hi @davecivicrm,

Thanks for advice, have refactored the code and added test. Also covered the scenarios when creating a contribution with cancelled/refunded status to keep consistency.

Best

Guanhuan

File: CRM/Contribute/BAO/Contribution.php [hunk 1]
```diff
@@ -3729,4 +3763,31 @@ public static function checkFinancialTypeChange($financialTypeId, $contributionI
...
 
+  /**
+   * Generate credit note id with next avaible number 
+   *
+   * @return string
+   *   Credit Note Id.
+   */
+  public static function createCreditNoteId() {
+    $prefixValue = CRM_Core_BAO_Setting::getItem(CRM_Core_BAO_Setting::CONTRIBUTE_PREFERENCES_NAME, 'contribution_invoice_settings');
+
```
Comment (davecivicrm) [inline CRM/Contribute/BAO/Contribution.php]: I think next 4 lines can be condensed to:
$creditNoteNum = CRM_Core_DAO::singleValueQuery("SELECT count(creditnote_id) as creditnote_number FROM civicrm_contribution");

**Comment by davecivicrm** [comment]
@guanhuan This looks good now, and thanks for adding a test. I've made one more optimization comment. AND we need to wait for PR testing to come back online (hopefully tomorrow) - but we're almost there :-)

**Comment by guanhuan** [comment]
Hi @davecivicrm,

Few coding standard issues are fixed. All set. Please take a look and let me know.

Best

Guanhuan

**Comment by davecivicrm** [comment]
Thanks @guanhuan Looks good!

**Pull request merged**